File: 5_polymorphism.py
import math

# Base class
class Shape:
    # Virtual function for calculating area
    def calculate_area(self):
        return 0

# Derived class for rectangle
class Rectangle(Shape):
    def __init__(self, length, breadth):
        self.length = length
        self.breadth = breadth

    # Override the calculate_area method
    def calculate_area(self):
        return self.length * self.breadth

# Derived class for circle
class Circle(Shape):
    def __init__(self, radius):
        self.radius = radius

    # Override the calculate_area method
    def calculate_area(self):
        return math.pi * self.radius * self.radius

# Function that uses polymorphism to calculate area
def print_area(shape):
    print("Area:", shape.calculate_area())

def main():
    # Creating objects of different shapes
    rectangle = Rectangle(4, 6)
    circle = Circle(3.5)

    # Using polymorphic behavior
    print_area(rectangle)
    print_area(circle)

if __name__ == "__main__":
    main()

File: 5_polymorphism.cpp
// Polymorphism in C++ allows objects of different classes to be treated as objects of a common superclass.

#include<bits/stdc++.h>

using namespace std;

// Base class
class Shape {
public:
    // Virtual function for calculating area
    virtual float calculate_area() {
        return 0;
    }
};

// Derived class for rectangle
class Rectangle : public Shape {
private:
    float length;
    float breadth;

public:
    Rectangle(float l, float b) : length(l), breadth(b) {}

    // Override the calculate_area function
    float calculate_area() override {
        return length * breadth;
    }
};

// Derived class for circle
class Circle : public Shape {
private:
    float radius;

public:
    Circle(float r) : radius(r) {}

    // Override the calculate_area function
    float calculate_area() override {
        return 3.14f * radius * radius;
    }
};

// Function that uses polymorphism to calculate area
void print_area(Shape *shape) {
    std::cout << "Area: " << shape->calculate_area() << std::endl;
}

int main() {
    // Creating objects of different shapes
    Rectangle rectangle(4, 6);
    Circle circle(3.5);

    // Using polymorphic behavior
    print_area(&rectangle);
    print_area(&circle);

    return 0;
}
